src/mds/MDCache.cc

```diff
@@ -6469,13 +6451,59 @@ void MDCache::_truncate_inode(CInode *in, LogSegment *ls)
     snapc = &nullsnap;
     ceph_assert(in->last == CEPH_NOSNAP);
   }
-  dout(10) << "_truncate_inode  snapc " << snapc << " on " << *in << dendl;
+  dout(10) << "_truncate_inode  snapc " << snapc << " on " << *in
+           << " fscrypt_last_block length is " << pi->fscrypt_last_block.length()
+           << dendl;
   auto layout = pi->layout;
-  filer.truncate(in->ino(), &layout, *snapc,
-		 pi->truncate_size, pi->truncate_from-pi->truncate_size,
-		 pi->truncate_seq, ceph::real_time::min(), 0,
-		 new C_OnFinisher(new C_IO_MDC_TruncateFinish(this, in, ls),
-				  mds->finisher));
+  struct ceph_fscrypt_last_block_header header;
+  memset(&header, 0, sizeof(header));
+  bufferlist data;
+  if (pi->fscrypt_last_block.length()) {
+    auto bl = pi->fscrypt_last_block.cbegin();
+    DECODE_START(1, bl);
+    decode(header.objver, bl);
+    decode(header.file_offset, bl);
+    decode(header.block_size, bl);
+    bl.copy(header.block_size, data);
+    DECODE_FINISH(bl);
+  }
+
+  /*
+   * If the block_size is 0, that means the fscrypt_last_block is empty
+   * or there has no data need to write in fscrypt_last_block, which
+   * means the truncate size is located in the file hole.
+   */
+  if (header.block_size) {
+    dout(10) << "_truncate_inode write_trunc on inode " << *in << " objver: "
+             << header.objver << " offset: " << header.file_offset << " blen: "
+	     << header.block_size << dendl;
+    /*
+     * When the Rados receives a write_trunc() request, if the new truncate seq
+     * is larger than current one in Rados, it will assume a write request comes
+     * before the truncate request, and it will do the truncate first and then
+     * write the new data to it.
+     *
+     * Since the offset + blen should always equal to pi->truncate_size, so in
+     * Rados the new writing data end won't exceed the truncated size it just did.
+     *
+     * So the write_trunc() is enough, no need to do the extra filer.truncate().
+     */
+    dout(10) << "_truncate_inode write_trunc on inode " << *in << " objver: "
+             << header.objver << " offset: " << header.file_offset << " blen: "
+	     << header.block_size << dendl;
+    filer.write_trunc(in->ino(), &layout, *snapc, header.file_offset, header.block_size,
```

Comment:
The `write_trunc()` works well in case the `truncate_from - truncate_size < 4MB` case, in other case it will keep the stale contents after truncate to a smaller size then a bigger one.

Reply:
Nice catch! I'll plan to do some testing with this as well.